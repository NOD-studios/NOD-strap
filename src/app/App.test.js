import 'react-dom';
import nock from 'nock';
import React, { Component } from 'react';
import configureMockStore from 'redux-mock-store';
import ReactTestRenderer from 'react-test-renderer';
import { createEpicMiddleware } from 'redux-observable';
import {
  router,
  fetchIp,
  fetchIpReject,
  fetchIpResolve,
  fetchIpCancelled,
  postForm,
  postFormReject,
  postFormResolve,
  postFormCancelled,
  rootEpic,
  fetchIpEpic,
  postFormEpic
} from '../index';

describe('app', () => {

  describe('App', () => {

    let tree;

    it('can render', () => {
      tree = ReactTestRenderer.create(router());
    });

    it('matches snapshot', () => {
      expect(tree).toMatchSnapshot();
    });

  });

  describe('epics', () => {

    const
      logIfCouldNotNock = log => log.match(': false') ? console.warn(log) : () => {},
      epicMiddleware = createEpicMiddleware(rootEpic),
      mockStore = configureMockStore([epicMiddleware]),
      endpoint = nock('https://httpbin.org')
        .log(logIfCouldNotNock);

    let store;

    describe('fetchIpEpic', () => {

      const
        data = { origin : '127.0.0.1' },
        fetchIpAction = fetchIp();

      afterAll(() => {
        if (!endpoint.isDone()) {
          console.error('pending mocks: %j', endpoint.pendingMocks());
        }
      });

      beforeEach(() => {
        store = mockStore();
      });

      afterEach(() => {
        nock.cleanAll();
        epicMiddleware.replaceEpic(fetchIpEpic);
      });

      it('fetching ip', () => {
        endpoint
          .get('/ip')
          .reply(200, data);

        store.dispatch(fetchIpAction);


        expect(store.getActions()).toEqual([fetchIpAction]);
      });

      it('fetching ip resolved', () => {

        endpoint
          .get('/ip')
          .reply(200, data);

        const action = fetchIpResolve(data);

        store.dispatch(fetchIpAction);
        store.dispatch(action);

        expect(store.getActions()).toEqual([fetchIpAction, action]);
      });

      it('fetching ip cancelled', () => {

        endpoint
          .get('/ip')
          .reply(200, data);

        const action = fetchIpCancelled();

        store.dispatch(fetchIpAction);
        store.dispatch(action);

        expect(store.getActions()).toEqual([fetchIpAction, action]);
      });

      it('fetching ip rejected', () => {

        const error = new Error('testing');

        endpoint
          .get('/ip')
          .reply(404, error);

        const action = fetchIpReject(error);
        store.dispatch(fetchIpAction);
        store.dispatch(action);

        expect(store.getActions()).toEqual([fetchIpAction, action]);
      });

    });

    describe('postFormEpic', () => {

      const
        data = {
          username : 'test',
          password : 'test'
        },
        postFormAction = postForm(data);

      afterAll(() => {
        if (!endpoint.isDone()) {
          console.error('pending mocks: %j', endpoint.pendingMocks());
        }
      });

      beforeEach(() => {
        store = mockStore();
      });

      afterEach(() => {
        nock.cleanAll();
        epicMiddleware.replaceEpic(postFormEpic);
      });

      it('posting form', () => {

        endpoint
          .post('/post', JSON.stringify(data))
          .reply(201, {
            json : data
          });

        store.dispatch(postFormAction);
        expect(store.getActions()).toEqual([postFormAction]);
      });

      it('posting form resolved', () => {
        const action = postFormResolve(data, false);

        store.dispatch(postFormAction);
        store.dispatch(action);

        expect(store.getActions()).toEqual([postFormAction, action]);
      });

      it('posting form cancelled', () => {

        const action = postFormCancelled();

        store.dispatch(postFormAction);
        store.dispatch(action);

        expect(store.getActions()).toEqual([postFormAction, action]);
      });

      it('posting form rejected', () => {

        const error = new Error('testing');

        const action = postFormReject(error, false);
        store.dispatch(postFormAction);
        store.dispatch(action);

        expect(store.getActions()).toEqual([postFormAction, action]);
      });

    });

  });

});
