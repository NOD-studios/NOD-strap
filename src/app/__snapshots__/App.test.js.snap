exports[`app App matches snapshot 1`] = `
<div
  className="App">
  <div
    className="p-0 container-fluid">
    <div
      className="row">
      <div
        className="col-xs-12">
        <nav
          className="navbar navbar-light bg-faded"
          role="navigation">
          <a
            className="navbar-brand"
            href="/"
            onClick={[Function]}
            style={Object {}}>
            Seed
          </a>
          <ul
            className="float-xs-right nav navbar-nav">
            <li
              className="nav-item">
              <a
                className="nav-link"
                href="/welcome"
                onClick={[Function]}
                style={Object {}}>
                Welcome
              </a>
            </li>
            <li
              className="nav-item">
              <a
                className="nav-link"
                href="/register"
                onClick={[Function]}
                style={Object {}}>
                Register
              </a>
            </li>
            <li
              className="nav-item">
              <a
                className="nav-link"
                href="http://github.com/NOD-studios/seed"
                onClick={[Function]}
                target="_blank">
                GitHub
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
    <main
      role="main">
      <div
        className="p-2 main-middle flex-items-xs-middle row">
        
        <div
          className="col-xs-12 col-md-6 offset-md-3">
          <div
            className="text-xs-center jumbotron">
            <h1
              className="display-4">
              Welcome
            </h1>
            <div
              className="row">
              <div
                className="col-xs-12">
                Â 
              </div>
            </div>
            <span
              className={undefined}
              style={
                Object {
                  "animation": "__react-md-spinner-animation__root-rotate 1568.235294117647ms linear infinite",
                  "display": "inline-block",
                  "height": 28,
                  "position": "relative",
                  "verticalAlign": "middle",
                  "width": 28,
                }
              }>
              <span
                style={
                  Object {
                    "animationDuration": "5332ms",
                    "animationIterationCount": "infinite",
                    "animationName": "__react-md-spinner-animation__fill-unfill-rotate, __react-md-spinner-animation__layer-1-fade-in-out",
                    "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                    "borderColor": "rgb(66, 165, 245)",
                    "display": "flex",
                    "height": "100%",
                    "opacity": 1,
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "100%",
                  }
                }>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__left-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(66, 165, 245) transparent transparent rgb(66, 165, 245)",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": 0,
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__right-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(66, 165, 245) rgb(66, 165, 245) transparent transparent",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": "-100%",
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(-129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "borderRadius": "50%",
                      "borderTopStyle": "solid",
                      "borderWidth": 3,
                      "boxSizing": "border-box",
                      "left": "45%",
                      "position": "absolute",
                      "top": 0,
                      "width": "10%",
                    }
                  } />
              </span>
              <span
                style={
                  Object {
                    "animationDuration": "5332ms",
                    "animationIterationCount": "infinite",
                    "animationName": "__react-md-spinner-animation__fill-unfill-rotate, __react-md-spinner-animation__layer-2-fade-in-out",
                    "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                    "borderColor": "rgb(239, 83, 80)",
                    "display": "flex",
                    "height": "100%",
                    "opacity": 1,
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "100%",
                  }
                }>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__left-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(239, 83, 80) transparent transparent rgb(239, 83, 80)",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": 0,
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__right-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(239, 83, 80) rgb(239, 83, 80) transparent transparent",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": "-100%",
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(-129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "borderRadius": "50%",
                      "borderTopStyle": "solid",
                      "borderWidth": 3,
                      "boxSizing": "border-box",
                      "left": "45%",
                      "position": "absolute",
                      "top": 0,
                      "width": "10%",
                    }
                  } />
              </span>
              <span
                style={
                  Object {
                    "animationDuration": "5332ms",
                    "animationIterationCount": "infinite",
                    "animationName": "__react-md-spinner-animation__fill-unfill-rotate, __react-md-spinner-animation__layer-3-fade-in-out",
                    "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                    "borderColor": "rgb(253, 216, 53)",
                    "display": "flex",
                    "height": "100%",
                    "opacity": 1,
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "100%",
                  }
                }>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__left-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(253, 216, 53) transparent transparent rgb(253, 216, 53)",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": 0,
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__right-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(253, 216, 53) rgb(253, 216, 53) transparent transparent",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": "-100%",
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(-129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "borderRadius": "50%",
                      "borderTopStyle": "solid",
                      "borderWidth": 3,
                      "boxSizing": "border-box",
                      "left": "45%",
                      "position": "absolute",
                      "top": 0,
                      "width": "10%",
                    }
                  } />
              </span>
              <span
                style={
                  Object {
                    "animationDuration": "5332ms",
                    "animationIterationCount": "infinite",
                    "animationName": "__react-md-spinner-animation__fill-unfill-rotate, __react-md-spinner-animation__layer-4-fade-in-out",
                    "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                    "borderColor": "rgb(76, 175, 80)",
                    "display": "flex",
                    "height": "100%",
                    "opacity": 1,
                    "position": "absolute",
                    "whiteSpace": "nowrap",
                    "width": "100%",
                  }
                }>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__left-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(76, 175, 80) transparent transparent rgb(76, 175, 80)",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": 0,
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "display": "inline-block",
                      "flexGrow": 1,
                      "height": "100%",
                      "overflow": "hidden",
                      "position": "relative",
                    }
                  }>
                  <span
                    style={
                      Object {
                        "animationDuration": "1333ms",
                        "animationIterationCount": "infinite",
                        "animationName": "__react-md-spinner-animation__right-spin",
                        "animationTimingFunction": "cubic-bezier(.4, 0, .2, 1)",
                        "borderColor": "rgb(76, 175, 80) rgb(76, 175, 80) transparent transparent",
                        "borderRadius": "50%",
                        "borderStyle": "solid",
                        "borderWidth": 3,
                        "bottom": 0,
                        "boxSizing": "border-box",
                        "left": "-100%",
                        "position": "absolute",
                        "top": 0,
                        "transform": "rotate(-129deg)",
                        "width": "200%",
                      }
                    } />
                </span>
                <span
                  style={
                    Object {
                      "borderColor": "inherit",
                      "borderRadius": "50%",
                      "borderTopStyle": "solid",
                      "borderWidth": 3,
                      "boxSizing": "border-box",
                      "left": "45%",
                      "position": "absolute",
                      "top": 0,
                      "width": "10%",
                    }
                  } />
              </span>
            </span>
          </div>
        </div>
      </div>
    </main>
    <div
      className="row">
      <div
        className="col-xs-12">
        <footer
          className="text-xs-center">
          <p>
            <small>
              Â© 
              2016
            </small>
          </p>
        </footer>
      </div>
    </div>
  </div>
</div>
`;
